consensus/consensus-types/src/block_retrieval.rs

```diff
@@ -161,28 +197,55 @@ impl BlockRetrievalResponse {
         &self.blocks
     }
 
-    /// TODO @bchocho @hariria change `retrieval_request` after all nodes upgrade to release with enum BlockRetrievalRequest (not struct)
+    pub fn verify_inner(&self, retrieval_request: &BlockRetrievalRequest) -> anyhow::Result<()> {
+        match &retrieval_request {
+            BlockRetrievalRequest::V1(retrieval_request) => {
+                ensure!(
+                    self.status != BlockRetrievalStatus::Succeeded
+                        || self.blocks.len() as u64 == retrieval_request.num_blocks(),
+                    "not enough blocks returned, expect {}, get {}",
+                    retrieval_request.num_blocks(),
+                    self.blocks.len(),
+                );
+                ensure!(
+                    self.status != BlockRetrievalStatus::SucceededWithTarget
+                        || self
+                            .blocks
+                            .last()
+                            .map_or(false, |block| retrieval_request.match_target_id(block.id())),
+                    "target not found in blocks returned, expect {:?}",
+                    retrieval_request.target_block_id(),
+                );
+            },
+            BlockRetrievalRequest::V2(retrieval_request) => {
+                ensure!(
+                    self.status != BlockRetrievalStatus::Succeeded
```

Comment:
also I'm not sure if this is robust enough towards malicious responses, if it should be succeed with target but instead I return a succeeded etc.

Reply:
Yeah it does seem this case is troublesome: the malicious node returns valid blocks that should be SucceededWithTarget, but only returns status as Succeeded. Is this what you're thinking of?